datafusion/core/src/physical_plan/sorts/sort.rs

```diff
@@ -253,11 +251,27 @@ impl ExternalSorter {
     /// Appends an unsorted [`RecordBatch`] to `in_mem_batches`
     ///
     /// Updates memory usage metrics, and possibly triggers spilling to disk
-    async fn insert_batch(&mut self, input: RecordBatch) -> Result<()> {
+    async fn insert_batch(&mut self, mut input: RecordBatch) -> Result<()> {
         if input.num_rows() == 0 {
             return Ok(());
         }
 
+        let mut batch_sorted = false;
+        if self.fetch.map_or(false, |f| f < input.num_rows()) {
+            // Eagerly sort the batch to potentially reduce the number of rows
+            // after applying the fetch parameter; first perform a memory reservation
+            // for the sorting procedure.
+            let mut reservation =
```

Comment:
I think the reasoning for the new consumer was that we only want to reserve a bit of memory briefly to account for the overhead of keeping both the sorted and original batch at the same time. Given that in this case we drop the old batch asap (due to `input` re-assignment) in favor of a smaller/truncated one I'd agree that a new consumer is not needed in that case. (Same could be said about the consumer/reservation in `sort_batch_stream`,  though in that case since it isn't given that there is a `LIMIT` we'd end up holding 2 same-ish batches at one point in time.)

That said, `reservation.try_grow(input.get_array_memory_size())` does then get called immediately below to check whether the new batch can be kept in memory without breaking the configured limit, so I don't think there's a need for another `try_grow` prior to that.

Reply:
Ah that makes sense

I think I think you could use (they very newly added) `Reservation::split()` to track the short allocation and free it on drop without needing an entirely new consumer:

https://github.com/apache/arrow-datafusion/blob/23547587c2773ddddb9b16cba1eb8ebf2eebd85a/datafusion/execution/src/memory_pool/mod.rs#L225-L231

perhaps like 
```rust
let batch_reservation = self.reservation.split(input.get_arra_memory_size());
input = sort_batch(&input, &self.expr, self.fetch)?;
// free allocation for previous input
// (it would also be freed on drop so the free isn't necessary but may be clearer)
batch_reservation.free();
```

Reply:
So I went ahead to add this, but then it seemed to me that the logic of `split` (and `take` by extension) is backwards? In particular, the `size` of the reservation is not _free_ memory, but instead _used_ memory, right? And with `split` we'd effectively deduct some piece of memory in our accounting that was previously allocated (without actually freeing that memory). 

Atm I went with skipping reservation updates at all in this case; the reason being that we're already holding the input batch in memory and we're are about to (try to) allocate memory for it right after sorting it anyway. Hence, the sole purpose of that reservation seems to be to account for that extra piece of memory due to non-in-place sorting (i.e. having the old and the new bath present at the same time briefly), and that piece of memory is guaranteed to be less than the batch size. Let me know if you'd like me to do something else.

I also changed that tests output. Now all tests should be passing, so it's probably time for the regular benchmarks.

Reply:
> Atm I went with skipping reservation updates at all in this case; 

I think this make sense -- especially since the extra memory for batches is likely small given k is very often small.